_s functions on all unsafe funtions

nrathaus opened the pull request:
Hi,

My pull request contains two main things:
1. Support for BSM files (its an XML standard we use for our testing engine)
2. Conversion of all unsafe functions to their safe counterparts (sprintf, fopen, strcat)

You may want to merge the second part at the very least

howerj commented:
Hi nrathaus,

Thanks for the work, however there are some problems. The '_s' are only available on Windows (I think some are defined in C11, but not all of them), they are not available on Linux, so I cannot merge these changes (I can't build the merged results). The code is currently written purely in C99, using functionality only available in the C standard library (or should be, if it is not then this is a bug). I believe there is a macro you can define (_CRT_SECURE_NO_WARNINGS) which should suppress the warnings.

As for the other changes relating to the new file format. The formatting is incorrect (well, different, not incorrect). The rest of the project uses tabs for indentation. :)

A few things we might want to think about before merging this is (I would like your feedback on these, feel free to argue against these points):

* The dbcc already generates an XML file, it would be better to use an XSLT file to transform the results instead of creating a new generator. I know there are other, XML based formats, for describing CAN data, instead of creating a new XML generator in C for them, an XSLT would be less error error prone. (XSLT is a bit of a pain, so it might be easier not to do this, even if it is not optimal).
* I would prefer that the new format if it were to be integrated would go into a separate file, I cannot verify that the results are correct because I have no way of verifying if my changes break the new format or not. I know this would duplicate some functionality in '2xml.c', but it mean I would accidentally break something if I modified the XML generators behaviour.
* A way I could verify the new format is with an XSD file, does one exist for the beSTORM format?
I have created an XSD and and XSLT for reference that work with the generated output for reference, I might be able to work on them, but I am quite busy at the moment.

Thanks,
Richard

nrathaus commented:
Hi,



On Thu, Nov 15, 2018 at 10:51 PM Richard James Howe <
notifications@github.com> wrote:

> Hi nrathaus,
>
> Thanks for the work, however there are some problems. The '_s' are only
> available on Windows (I think some are defined in C11, but not all of
> them), they are not available on Linux, so I cannot merge these changes (I
> can't build the merged results). The code is currently written purely in
> C99, using functionality only available in the C standard library (or
> should be, if it is not then this is a bug). I believe there is a macro you
> can define (_CRT_SECURE_NO_WARNINGS) which should suppress the warnings.
>
I believe I went with the C11 changes (safe functions), I understand if you
don't want to merge them, I think in 2018, using strcat, strcpy, sprintf,
and similar without doing length checking is a bit of a taboo - but its
your project and your call to make


> As for the other changes relating to the new file format. The formatting
> is incorrect (well, different, not incorrect). The rest of the project uses
> tabs for indentation. :)
>

That is Visual Studio, using spaces rather than tabs - I will revert that
change


> A few things we might want to think about before merging this is (I would
> like your feedback on these, feel free to argue against these points):
>
>    - The dbcc already generates an XML file, it would be better to use an
>    XSLT file to transform the results instead of creating a new generator. I
>    know there are other, XML based formats, for describing CAN data, instead
>    of creating a new XML generator in C for them, an XSLT would be less error
>    error prone. (XSLT is a bit of a pain, so it might be easier not to do
>    this, even if it is not optimal).
>
>
XSLT is a good path, but would require a XSLT parser/processor - an added
step that would complicate our use of your program


>
>    - I would prefer that the new format if it were to be integrated would
>    go into a separate file, I cannot verify that the results are correct
>    because I have no way of verifying if my changes break the new format or
>    not. I know this would duplicate some functionality in '2xml.c', but it
>    mean I would accidentally break something if I modified the XML generators
>    behaviour.
>
>
No problem, I will split the 2xml.c to 2bsm.c so that its easier to manage


>
>    - A way I could verify the new format is with an XSD file, does one
>    exist for the beSTORM format?
>    I have created an XSD and and XSLT for reference that work with the
>    generated output for reference, I might be able to work on them, but I am
>    quite busy at the moment.
>
>
We don't have XSD, we do have a DTD (attached)


>
>
> Thanks,
> Richard
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/howerj/dbcc/pull/14#issuecomment-439185534>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB8CbsKCsdrSw93vyO230pNA_lvf0w75ks5uvdPOgaJpZM4YfLe_>
> .
>


-- 

Thanks,
Noam Rathaus
Beyond Security



PGP Key ID: 7EF920D3C045D63F (Exp 2019-03)

nrathaus commented:
Hi,

I removed the BSM code from 2xml and changed all the spaces into tabs

If you wish you could merge it, or not, I won't mind


On Sun, Nov 18, 2018 at 9:14 AM Noam Rathaus <noamr@beyondsecurity.com>
wrote:

> Hi,
>
>
>
> On Thu, Nov 15, 2018 at 10:51 PM Richard James Howe <
> notifications@github.com> wrote:
>
>> Hi nrathaus,
>>
>> Thanks for the work, however there are some problems. The '_s' are only
>> available on Windows (I think some are defined in C11, but not all of
>> them), they are not available on Linux, so I cannot merge these changes (I
>> can't build the merged results). The code is currently written purely in
>> C99, using functionality only available in the C standard library (or
>> should be, if it is not then this is a bug). I believe there is a macro you
>> can define (_CRT_SECURE_NO_WARNINGS) which should suppress the warnings.
>>
> I believe I went with the C11 changes (safe functions), I understand if
> you don't want to merge them, I think in 2018, using strcat, strcpy,
> sprintf, and similar without doing length checking is a bit of a taboo -
> but its your project and your call to make
>
>
>> As for the other changes relating to the new file format. The formatting
>> is incorrect (well, different, not incorrect). The rest of the project uses
>> tabs for indentation. :)
>>
>
> That is Visual Studio, using spaces rather than tabs - I will revert that
> change
>
>
>> A few things we might want to think about before merging this is (I would
>> like your feedback on these, feel free to argue against these points):
>>
>>    - The dbcc already generates an XML file, it would be better to use
>>    an XSLT file to transform the results instead of creating a new generator.
>>    I know there are other, XML based formats, for describing CAN data, instead
>>    of creating a new XML generator in C for them, an XSLT would be less error
>>    error prone. (XSLT is a bit of a pain, so it might be easier not to do
>>    this, even if it is not optimal).
>>
>>
> XSLT is a good path, but would require a XSLT parser/processor - an added
> step that would complicate our use of your program
>
>
>>
>>    - I would prefer that the new format if it were to be integrated
>>    would go into a separate file, I cannot verify that the results are correct
>>    because I have no way of verifying if my changes break the new format or
>>    not. I know this would duplicate some functionality in '2xml.c', but it
>>    mean I would accidentally break something if I modified the XML generators
>>    behaviour.
>>
>>
> No problem, I will split the 2xml.c to 2bsm.c so that its easier to manage
>
>
>>
>>    - A way I could verify the new format is with an XSD file, does one
>>    exist for the beSTORM format?
>>    I have created an XSD and and XSLT for reference that work with the
>>    generated output for reference, I might be able to work on them, but I am
>>    quite busy at the moment.
>>
>>
> We don't have XSD, we do have a DTD (attached)
>
>
>>
>>
>> Thanks,
>> Richard
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/howerj/dbcc/pull/14#issuecomment-439185534>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AB8CbsKCsdrSw93vyO230pNA_lvf0w75ks5uvdPOgaJpZM4YfLe_>
>> .
>>
>
>
> --
>
> Thanks,
> Noam Rathaus
> Beyond Security
>
>
>
> PGP Key ID: 7EF920D3C045D63F (Exp 2019-03)
>
>

-- 

Thanks,
Noam Rathaus
Beyond Security



PGP Key ID: 7EF920D3C045D63F (Exp 2019-03)

howerj commented:
Cheers, I've merged that, but I removed the '_s' functions purely for portability reasons, C11 Annex K is optional after all. I do agree in principle that there should be more length checking, although I do not think Annex K is the solution.

I noticed you also sorted the signals on the CAN ID, all the output formats look a lot neater and organized. Thanks.

nrathaus commented:
Hi

1.  No problem about the _s getting dumped - it’s my own feel that security
is a default these days - but I understand that portability is also
important

I am not aware of any easy way to _s style functions without using c11 -
doing all the length checks manually would make the code complex and error
prone

2. I sorted it because I found it harder to debug when it didn’t work -
happy you liked it

3. Notice that the code also defects gaps in the signal analysis- and will
mark it as UNKNOWN- this is also helpful for the users when they try to
understand why he isn’t getting anything for bits 8-10 (for example)

On Tue, 20 Nov 2018 at 00:12, Richard James Howe <notifications@github.com>
wrote:

> Cheers, I've merged that, but I removed the '_s' functions purely for
> portability reasons, C11 Annex K is optional after all. I do agree in
> principle that there should be more length checking, although I do not
> think Annex K is the solution.
>
> I noticed you also sorted the signals on the CAN ID, all the output
> formats look a lot neater and organized. Thanks.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/howerj/dbcc/pull/14#issuecomment-440060477>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB8CbgOClOaKhyzIHL5dsEPGn4ShUi-Gks5uwyzJgaJpZM4YfLe_>
> .
>
-- 
Thanks,
Noam Rathaus
